### DISCUSSION TITLE
Mental model for migrations

### DISCUSSION THREAD

**Discussion by einarpersson** [pocketbase/pocketbase#1947] [Q&A] [answered]
Imagine I have a `pb_migrations`-folder with a bunch of files, and I have a clean `pb_data`.

1. When I run `pocketbase serve` my understanding is that these migration files will be applied automatically, is this correct? I note that there is a `--automigrate` global flag but it is not 100% clear if this applies to the automatic _creation_ of migration files or _running_ of migration files (or both). Which is it?
2. If I were to manually run a migration one by one, are the applied migrations stored somewhere in the database? Is there a hidden migrations table which can be inspected somehow? I have right now a situation in my local dev db where it says that there are no new migrations to apply, but if I try to migrate down it lists an older migrations file as the last one. Since this is a dev setup I could and will probably just delete `pb_data` but it would be interesting to understand what is going wrong.
3. During development it is easy to produce a large number migration files since the schema is volatile. Once the schema has begun to settle, can / should I clean `pb_migrations` and use the `pocketbase migrate collections` to create a new starting point? But will this work with the production db which has already had a number of "older migrations" applied to it? I mean, you can arrive at the same schema by running 14 migrations or just one. If there is a live db that has had 14 migrations applied, can I somehow seamlessly "replace that history" so that there is just one migrations file and build on from there? (adding a second one, a third one, etc).

I'm sorry if it feels as if I'm mixing several questions, but the overall theme is a lacking mental model of migrations. Thankful for any comments that can bring clarity.

I would be glad to contribute to the [docs](https://github.com/pocketbase/site/blob/master/src/routes/(app)/docs/migrations/%2Bpage.svelte) with minor additions once I get a better understanding myself. I know that you've previously stated that you want to make an overhaul of the js migrations docs, but in the meantime I think that minor things could go a long way for new users.

**Comment by ganigeorgiev** [comment, marked as answer]
> When I run pocketbase serve my understanding is that these migration files will be applied automatically, is this correct?

Yes. `serve` will run all unapplied migrations from both the `pb_migrations` directory and the ones that are embedded in the binary. The `--automigrate` flag is only for the automatic migration creation, for example when doing changes from the UI.

---

> If I were to manually run a migration one by one, are the applied migrations stored somewhere in the database? 

Yes. All applied migrations are stored in the `_migrations` table. `migrate down` will revert the last applied migration(s).

---

> During development it is easy to produce a large number migration files since the schema is volatile. Once the schema has begun to settle, can / should I clean pb_migrations and use the pocketbase migrate collections to create a new starting point? But will this work with the production db which has already had a number of "older migrations" applied to it?

It is safe to run `pocketbase migrate collections` as many times as you want. It creates a full snapshot of all collections configuration but keep in mind that if another developer has also created migration with a full snapshot, on merge you'll have to decide which version to keep since the last snapshot state will overwrite any previous one.

PocketBase migrations flow is similar to the other common frameworks/orms that you can find in Laravel, Django, etc.

If you delete the old migration files your app will still continue to work, but you'll not be able to revert beyond the point of the last deleted previously applied migration.

---

> If there is a live db that has had 14 migrations applied, can I somehow seamlessly "replace that history" so that there is just one migrations file and build on from there?

There is no clean/replace command at the moment and you'll have to delete manually the entries from the `_migrations` table if you want to do that. I've added it to the [roadmap](https://github.com/orgs/pocketbase/projects/2) to consider implementing a "migrate clean/sync" or similar command in the next release that will delete all missing previously applied migrations from the `_migrations` table.

**Reply by ganigeorgiev** [reply]
_Just for clarity, if you are interested in the implementation details how migrations are applied, you could check [`tools/migrate/runner.go`](https://github.com/pocketbase/pocketbase/blob/master/tools/migrate/runner.go)._
